src/Common/MemoryTracker.cpp

```diff
@@ -151,24 +162,6 @@ void MemoryTracker::allocImpl(Int64 size, bool throw_if_memory_exceeded, MemoryT
     Int64 current_hard_limit = hard_limit.load(std::memory_order_relaxed);
     Int64 current_profiler_limit = profiler_limit.load(std::memory_order_relaxed);
 
-    /// Cap the limit to the total_memory_tracker, since it may include some drift
-    /// for user-level memory tracker.
-    ///
-    /// And since total_memory_tracker is reset to the process resident
-    /// memory peridically (in AsynchronousMetrics::update()), any limit can be
-    /// capped to it, to avoid possible drift.
-    if (unlikely(current_hard_limit
-        && will_be > current_hard_limit
-        && level == VariableContext::User))
-    {
-        Int64 total_amount = total_memory_tracker.get();
-        if (amount > total_amount)
-        {
-            set(total_amount);
-            will_be = size + total_amount;
-        }
-    }
```

Comment:
Without this code queries with `max_memory_usage_for_user` may fail (#12182), if there will be significant drift in memory tracker, although AFAIR nowadays it should not have significant drift.

Reply:
I would rather find the reason for this kind of drift.
I hope that with current change
* memory for total_memory_tracker should be tracked precisely (except memory which was allocated with malloc inside some external libraries)
*  memory for other memory trackers should be almost precise (except while using MemoryTrackerBlockerInThread)
* RSS should be >= total_memory_tracker (because of untracked mallocs and MADV_FREE inside jemalloc)